packages/auth/src/mfa/assertions/totp.ts

```diff
@@ -0,0 +1,276 @@
+    return finalizeSignInTotpMfa(auth, {
+      mfaPendingCredential,
+      mfaEnrollmentId: this.enrollmentId,
+      totpVerificationInfo
+    });
+  }
+}
+
+/**
+ * Provider for generating a {@link TotpMultiFactorAssertion}.
+ *
+ * Stores the shared secret key and other parameters to generate time-based OTPs.
+ * Implements methods to retrieve the shared secret key, generate a QRCode URL.
+ * @public
+ */
+export class TotpSecret {
+  /**
+   * Shared secret key/seed used for enrolling in TOTP MFA and generating OTPs.
+   */
+  readonly secretKey: string;
+  /**
+   * Hashing algorithm used.
+   */
+  readonly hashingAlgorithm: string;
+  /**
+   * Length of the one-time passwords to be generated.
+   */
+  readonly codeLength: number;
+  /**
+   * The interval (in seconds) when the OTP codes should change.
+   */
+  readonly codeIntervalSeconds: number;
+  /**
+   * The timestamp(UTC string) by which TOTP enrollment should be completed.
+   */
+  // This can be used by callers to show a countdown of when to enter OTP code by.
+  readonly enrollmentCompletionDeadline: string;
+
+  // The public members are declared outside the constructor so the docs can be generated.
+  private constructor(
+    secretKey: string,
+    hashingAlgorithm: string,
+    codeLength: number,
+    codeIntervalSeconds: number,
+    enrollmentCompletionDeadline: string,
+    private readonly sessionInfo: string,
+    private readonly auth: AuthInternal
+  ) {
+    this.secretKey = secretKey;
+    this.hashingAlgorithm = hashingAlgorithm;
+    this.codeLength = codeLength;
+    this.codeIntervalSeconds = codeIntervalSeconds;
+    this.enrollmentCompletionDeadline = enrollmentCompletionDeadline;
+  }
+
+  /** @internal */
+  static _fromStartTotpMfaEnrollmentResponse(
+    response: StartTotpMfaEnrollmentResponse,
+    auth: AuthInternal
+  ): TotpSecret {
+    return new TotpSecret(
+      response.totpSessionInfo.sharedSecretKey,
+      response.totpSessionInfo.hashingAlgorithm,
+      response.totpSessionInfo.verificationCodeLength,
+      response.totpSessionInfo.periodSec,
+      new Date(response.totpSessionInfo.finalizeEnrollmentTime).toUTCString(),
+      response.totpSessionInfo.sessionInfo,
+      auth
+    );
+  }
+
+  /** @internal */
+  _makeTotpVerificationInfo(otp: string): TotpVerificationInfo {
+    return { sessionInfo: this.sessionInfo, verificationCode: otp };
+  }
+
+  /**
+   * Returns a QRCode URL as described in
+   * https://github.com/google/google-authenticator/wiki/Key-Uri-Format
+   * This can be displayed to the user as a QRCode to be scanned into a TOTP App like Google Authenticator.
+   * If the optional parameters are unspecified, an accountName of <userEmail> and issuer of <firebaseAppName> are used.
+   *
+   * @param accountName the name of the account/app along with a user identifier.
+   * @param issuer issuer of the TOTP(likely the app name).
+   * @returns A QRCode URL string.
+   */
```

Comment:
```suggestion
  /**
   * Returns a QR code URL as described in
   * https://github.com/google/google-authenticator/wiki/Key-Uri-Format
   * This can be displayed to the user as a QR code to be scanned into a TOTP app like Google Authenticator.
   * If the optional parameters are unspecified, an accountName of <userEmail> and issuer of <firebaseAppName> are used.
   *
   * @param accountName the name of the account/app along with a user identifier.
   * @param issuer issuer of the TOTP (likely the app name).
   * @returns A QR code URL string.
   */
```